src/uuid.cr

```diff
@@ -176,6 +211,121 @@ struct UUID
+  #   The clock sequence should be stored in a stable location, such as a file. If it is not
+  #   stored, a random value is used by default. If not provided the current time milliseconds
+  #   are used.
+  # * `node_id`: In case the traditional MAC address based approach should be taken the
+  #   `node_id` can be provided. Otherwise secure random is used.
+  def self.v1(*, clock_seq : UInt16? = nil, node_id : MAC? = nil) : self
+    tl = Time.local
+    now = (tl.to_unix_ns / 100).to_u64 + 122192928000000000
+    seq = ((clock_seq || (tl.nanosecond/1000000).to_u16) & 0x3fff) | 0x8000
+
+    time_low = UInt32.new(now & 0xffffffff)
+    time_mid = UInt16.new((now >> 32) & 0xffff)
+    time_hi = UInt16.new((now >> 48) & 0x0fff)
+    time_hi |= 0x1000 # Version 1
+
+    uuid = uninitialized UInt8[16]
+    IO::ByteFormat::BigEndian.encode(time_low, uuid.to_slice[0..3])
+    IO::ByteFormat::BigEndian.encode(time_mid, uuid.to_slice[4..5])
+    IO::ByteFormat::BigEndian.encode(time_hi, uuid.to_slice[6..7])
+    IO::ByteFormat::BigEndian.encode(seq, uuid.to_slice[8..9])
+
+    if node_id
+      6.times do |i|
+        uuid.to_slice[10 + i] = node_id[i]
+      end
+    else
+      Random::Secure.random_bytes(uuid.to_slice[10..15])
+      # set multicast bit as recommended per section 4.5 of the RFC 4122 spec
+      # to not conflict with real MAC addresses
+      uuid[10] |= 0x01_u8
+    end
+
+    new(uuid, version: UUID::Version::V1, variant: UUID::Variant::RFC4122)
+  end
+
+  # Generates RFC 4122 v2 UUID.
+  #
+  # Version 2 UUIDs are generated using the current time, the local machine’s MAC address,
+  # and the local user or group ID. However, they are not widely used due to their limitations.
+  # For a given domain/id pair, the same token may be returned for a duration of up to 7 minutes
+  # and 10 seconds.
+  #
+  # * `domain`: The `Domain` of the uuid.
+  # * `id`: The id depends on the domain, for the `Domain::Person` usually the local user id
+  #    is used, for `Domain::Group` usually the local group id is used.
+  # * `node_id`: In case the traditional MAC address based approach should be taken the
+  #   `node_id` can be provided. Otherwise secure random is used.
+  def self.v2(domain : Domain, id : UInt32, node_id : MAC? = nil) : self
+    uuid = v1(node_id: node_id).bytes
+    uuid[6] = (uuid[6] & 0x0f) | 0x20 # Version 2
+    uuid[9] = domain.to_u8
+    IO::ByteFormat::BigEndian.encode(id, uuid.to_slice[0..3])
+    new(uuid, version: UUID::Version::V2, variant: UUID::Variant::RFC4122)
+  end
+
+  # Generates RFC 4122 v3 UUID.
+  #
+  # * `name`: The name used to generate the UUID, it can be a string of any size.
+  # * `namespace`: Specifies the type of the name, usually one of `Namespace`
+  def self.v3(name : String, namespace : UUID) : self
+    klass = {% if flag?(:without_openssl) %}::Crystal::Digest::MD5{% else %}::Digest::MD5{% end %}
```

Comment:
Perhaps a bit cleaner:

```suggestion
    klass = {{ flag?(:without_openssl) ? "::Crystal::Digest::MD5".id : "::Digest::MD5".id }}
```

Reply:
I don't think this is any better.

Reply:
It would be better to move this logic to the "digest/sha1" so that any user in the stdlib would not have to do that. This way it would be more DRY. What do you think, should I also update this in this PR or keep it as is and add  another PR that focuses on that?

Reply:
I performed this micro benchmark:

```crystal
require "benchmark"
require "crystal/digest/sha1"
require "crystal/digest/md5"
require "digest/sha1"
require "digest/md5"
require "uuid"

namespace = UUID.new("6ba7b811-9dad-11d1-80b4-00c04fd430c8")
name = "https://crystal-lang.org/api/1.9.2/Benchmark.html"

Benchmark.ips do |x|
  x.report("md5 openssl") do
    klass = ::Digest::MD5
    hash = klass.digest do |ctx|
      ctx.update namespace.bytes
      ctx.update name
    end
  end
  x.report("md5 native") do
    klass = ::Crystal::Digest::MD5
    hash = klass.digest do |ctx|
      ctx.update namespace.bytes
      ctx.update name
    end
  end
end

Benchmark.ips do |x|
  x.report("sha1 openssl") do
    klass = ::Digest::SHA1
    hash = klass.digest do |ctx|
      ctx.update namespace.bytes
      ctx.update name
    end
  end
  x.report("sha1 native") do
    klass =::Crystal::Digest::SHA1
    hash = klass.digest do |ctx|
      ctx.update namespace.bytes
      ctx.update name
    end
  end
end
```

And got the following results: 

```
==. Crystal 1.9.2 [1908c816f] (2023-07-19) LLVM: 15.0.7
Default target: x86_64-unknown-linux-gnu
md5 openssl  940.28k (  1.06µs) (± 2.64%)  128B/op   2.08× slower
md5 native.    1.95M (512.47ns) (± 1.00%)  144B/op        fastest
sha1 openssl 945.85k (  1.06µs) (± 2.30%)  128B/op   1.76× slower
sha1 native    1.67M (599.63ns) (± 1.01%)  144B/op        fastest

== Crystal 1.9.2 (2023-07-19) LLVM: 15.0.7
Default target: aarch64-apple-darwin22.5.0
md5 openssl    2.34M (426.48ns) (± 0.78%)  128B/op   1.39× slower
md5 native     3.27M (305.99ns) (± 0.89%)  144B/op        fastest
sha1 openssl   2.93M (341.20ns) (± 1.03%)  128B/op   1.25× slower
sha1 native    3.66M (273.55ns) (± 0.37%)  144B/op        fastest
```

By this results there is seemingly no benefit of using the OpenSSL based SHA1 and MD5 variants for this workload. 
What do you think about removing the switch and go with the native one only?

Reply:
I removed the openssl code.

Reply:
> By this results there is seemingly no benefit of using the OpenSSL based SHA1 and MD5 variants for this workload.
> What do you think about removing the switch and go with the native one only?

There is a huge benefit of using the OpenSSL implementation: It's cryptographically more robust and audited (hopefully). The native Crystal implementation is not verified in any way. We should use OpenSSL unless explicitly opted out.

Reply:
Did as requested.